src/DCB/EventRouter.php

```diff
@@ -0,0 +1,172 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Patchlevel\EventSourcing\DCB;
+
+use Patchlevel\EventSourcing\Attribute\Apply;
+use Patchlevel\EventSourcing\Message\Message;
+use Patchlevel\EventSourcing\Store\SubQuery;
+use ReflectionClass;
+use ReflectionIntersectionType;
+use ReflectionMethod;
+use ReflectionNamedType;
+use ReflectionUnionType;
+
+use function array_key_exists;
+use function array_keys;
+use function array_map;
+use function array_merge;
+use function class_exists;
+
+/** @experimental */
+trait EventRouter
```

Comment:
I think this might be misleading from the naming pov, as this is not really a router implementation in a classical way. Here we have mixed in apply logic (maybe this can be called routing) and query building / filtering. That being said: I come not with a better name 🤣 

Maybe we should add a `require-implements` for the `Projection` interface here.
Ref: https://phpstan.org/writing-php-code/phpdocs-basics#enforcing-implementing-an-interface-for-traits